File: tree/leetcode_Binary_Tree_Maximum_Path_Sum.java
/*Binary_Tree_Maximum_Path_Sum

Given a binary tree, find the maximum path sum.
The path may start and end at any node in the tree.
For example:
Given the below binary tree,
    1
   / \
  2   3
Return 6.*/

public class leetcode_Binary_Tree_Maximum_Path_Sum {

    public static void main(String[] args) {
        Solution test = new Solution(); 
    }
}


class Solution {
    private int max;

    public int maxPathSum(TreeNode root) {
        this.max = root.val;
        nextPath(root);
        return this.max;
    }

    public int nextPath(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = nextPath(root.left);
        int right = nextPath(root.right);
        int localMax = root.val;
        if (left > 0) {
            localMax += left;
        }
        if (right > 0) {
            localMax += right;
        }
        this.max = Math.max(this.max, localMax);
        return Math.max(root.val, Math.max(root.val + left, root.val + right));
    }
}


class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;

    TreeNode(int val) {
		this.left = null;
		this.right = null;
        this.val = x;
    }
}

/* Python Version

class Solution:
    # @param root, a tree node
    # @return an integer
    def __init__(self):
        self.max_sum = None
    
    def maxPathSum(self, root):
        def postorder(root):
            if not root:
                return 0
            left = postorder(root.left)
            right = postorder(root.right)
            
            local_max = root.val
            
            if left > 0:
                local_max += left
            if right > 0:
                local_max += right
                
            self.max_sum = max(self.max_sum, local_max)
            return max(root.val, max(root.val + left, root.val + right))
        if not root:
            return 0

        self.max_sum = root.val
        postorder(root)
        return self.max_sum
*/

File: tree/leetcode_Binary_Tree_Maximum_Path_Sum.py
"""Binary tree maximum path sum
leetcode

Given a binary tree, find the maximum path sum.

For this problem, a path is defined as any sequence of nodes from some starting
node to any node in the tree along the parent-child connections. The path does
not need to go through the root.

For example:
Given the below binary tree,

       1
      / \
     2   3
Return 6.
"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.dfs(root)[1]

    def dfs(self, root):
        if root is None:
            return float('-inf'), float('-inf')

        left, left_max = self.dfs(root.left)
        right, right_max = self.dfs(root.right)

        local = root.val
        left_path = root.val
        right_path = root.val
        if left > 0:
            local += left
            left_path += left
        if right > 0:
            local += right
            right_path += right

        return max(left_path, right_path), max(left_max, right_max, local)
